File: src/data/courses/dsa/trie/trie-core-operations/count-operations/code/python/count_operations.py
# ============================================================
# Solution 1: Dict counting — O(N*L) prefix count
# ============================================================
from collections import Counter
class Solution1:
    def __init__(self): self.words = Counter()
    def insert(self, word): self.words[word] += 1
    def countWordsEqualTo(self, word): return self.words[word]
    def countWordsStartingWith(self, prefix):
        return sum(c for w, c in self.words.items() if w.startswith(prefix))


# ============================================================
# Solution 2: Optimal (Trie-based)
# ============================================================
class TrieNode:
    def __init__(self):
        self.children = {}  # Dictionary for child nodes
        self.word_count = 0  # Count of words ending here
        self.prefix_count = 0  # Count of words passing through

class CountOperations:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
            node.prefix_count += 1  # Increment prefix count
        node.word_count += 1  # Increment word count at end
    
    def countWordsEqualTo(self, word: str) -> int:
        node = self.root
        for c in word:
            if c not in node.children:
                return 0  # Word doesn't exist
            node = node.children[c]
        return node.word_count  # Return count of this exact word
    
    def countWordsStartingWith(self, prefix: str) -> int:
        node = self.root
        for c in prefix:
            if c not in node.children:
                return 0  # Prefix doesn't exist
            node = node.children[c]
        return node.prefix_count  # Return count of words with this prefix

File: src/data/courses/dsa/trie/trie-core-operations/count-operations/code/cpp/CountOperations.cpp
// ============================================================
// Solution 1: HashMap counting
// Time: O(1) word count, O(N*L) prefix count
// ============================================================
#include <unordered_map>
#include <string>
using namespace std;

class Solution1 {
    unordered_map<string, int> wordCount;
public:
    void insert(string word) { wordCount[word]++; }
    int countWordsEqualTo(string word) { return wordCount[word]; }
    int countWordsStartingWith(string prefix) {
        int count = 0;
        for (auto& [w, c] : wordCount) // O(N) scan all words!
            if (w.substr(0, prefix.size()) == prefix) count += c;
        return count;
    }
};

// ============================================================
// Solution 2: Optimal (Trie-based)
// ============================================================
#include <string>
using namespace std;

class CountOperations {
    struct TrieNode {
        TrieNode* children[26] = {}; // 26 lowercase letters
        int wordCount = 0; // Count of words ending here
        int prefixCount = 0; // Count of words passing through
    };
    
    TrieNode* root = new TrieNode();
    
public:
    void insert(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx]) {
                node->children[idx] = new TrieNode();
            }
            node = node->children[idx];
            node->prefixCount++; // Increment prefix count
        }
        node->wordCount++; // Increment word count at end
    }
    
    int countWordsEqualTo(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx]) return 0; // Word doesn't exist
            node = node->children[idx];
        }
        return node->wordCount; // Return count of this exact word
    }
    
    int countWordsStartingWith(string prefix) {
        TrieNode* node = root;
        for (char c : prefix) {
            int idx = c - 'a';
            if (!node->children[idx]) return 0; // Prefix doesn't exist
            node = node->children[idx];
        }
        return node->prefixCount; // Return count of words with this prefix
    }
};
